### ISSUE TITLE
This code fails when T is polymorphic

### ISSUE THREAD

**Issue by lifflander** [DARMA-tasking/checkpoint#268]
```C++
std::unique_ptr<T> task(new U(1234));
auto ret = checkpoint::serialize<T>(*task);
auto out = checkpoint::deserialize<T>(std::move(ret));
```

@PhilMiller What do you think? Should we support this case?

@nlslatt

**Comment by PhilMiller** [comment]
Yes, I think we should support this case. It presumably works fine if `task` is serialized as a member of something else, and it's just `s | task`?

**Comment by PhilMiller** [comment]
Since a PR for this just got posted, I'm looking at the example again, and I can't help but read that as kinda like slicing. Probably still good if it is made to work correctly, though. I'll look at the PR.
